components/org.wso2.carbon.identity.application.authenticator.push/src/main/java/org/wso2/carbon/identity/application/authenticator/push/PushAuthenticator.java

```diff
@@ -242,40 +254,164 @@ protected void processAuthenticationResponse(HttpServletRequest httpServletReque
+            throw new AuthenticationFailedException("Error occurred when trying to get the public key.");
+        } catch (PushDeviceHandlerClientException e) {
+            throw new AuthenticationFailedException("Public key error");
         }
-        return isValid;
+    }
+
+    /**
+     * Redirect user to device selection page.
+     *
+     * @param response          HTTP response
+     * @param context           Authentication context
+     * @param sessionDataKey    Unique key for the session
+     * @param deviceArrayString JSON array as a string
+     * @throws IOException if an error occurs when redirecting to the device selection page
+     */
+    private void redirectDevicesPage(HttpServletResponse response, AuthenticationContext context,
+                                     String sessionDataKey, String deviceArrayString) throws IOException {
+
+        Config config = new Config();
+        String devicesPage;
+        devicesPage = config.getDevicesPage(context)
+                + "?sessionDataKey=" + URLEncoder.encode(sessionDataKey, StandardCharsets.UTF_8.name())
+                + "&devices=" + URLEncoder.encode(deviceArrayString, StandardCharsets.UTF_8.name());
+        response.sendRedirect(devicesPage);
+    }
+
+    /**
+     * Redirect the user to the wait page.
+     *
+     * @param response       HTTP response
+     * @param sessionDataKey Unique ID for the session
+     * @param user           Authenticated user
+     * @throws PushAuthenticatorException if an error occurs while trying to redirect to the wait page
+     */
+    private void redirectWaitPage(HttpServletResponse response, String sessionDataKey, AuthenticatedUser user)
+            throws PushAuthenticatorException {
+
+        try {
+            String waitPage = PushAuthenticatorConstants.WAIT_PAGE
+                    + "?sessionDataKey="
+                    + URLEncoder.encode(sessionDataKey, StandardCharsets.UTF_8.name());
+            response.sendRedirect(waitPage);
+        } catch (IOException e) {
+            String errorMessage = String.format("Error occurred when trying to to redirect user: %s to the wait page.",
+                    user.toFullQualifiedUsername());
+            throw new PushAuthenticatorException(errorMessage, e);
+        }
+    }
+
+    /**
+     * Redirect the user to the authentication denied page.
+     *
+     * @param httpServletResponse HTTP response
+     * @throws IOException if an error occurs while redirecting to the denied page
+     */
+    private void redirectRetryPage(HttpServletResponse httpServletResponse, String status, String message)
+            throws IOException {
+
+        String deniedPage = "/authenticationendpoint/retry.do"
+                + "?status=" + URLEncoder.encode(status, StandardCharsets.UTF_8.name())
+                + "&statusMsg=" + URLEncoder.encode(message, StandardCharsets.UTF_8.name());
+        httpServletResponse.sendRedirect(deniedPage);
```

Comment:
Implement refering to the following https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1569/commits/07ad46babd5b6526efbc628e15eccc43e1789d72#diff-3f5f539c437be28f67a5c15372f267e1443763fc1d98a7de28453175f5661337R89-R90